decoder.go

```diff
@@ -0,0 +1,211 @@
+package redis
+
+import (
+	"errors"
+	"reflect"
+	"strconv"
+	"strings"
+	"sync"
+)
+
+// Decoder represents a type with a custom Decode() function that can
+// take a string and parse it to set a desired value on a struct field,
+// similar to how encoding/json implements UnmarshalJSON().
+type Decoder interface {
+	Decode(string) error
+}
+
+// defDecoder represents decoding functions for default built-in types.
+type defDecoder func(reflect.Value, string) error
+
+// structField represents a single field in a target struct.
+type structField struct {
+	index int
+
+	// Decoder.Decode() or defDecoder()
+	fn interface{}
```

Comment:
interface{} -> decoderFunc. To support `Decoder` interface we can later add:

```go
func decodeInterface(v reflect.Value, s string) error {
    return v.Interface().(Decoder).Decode(s)
}
```

But let's not worry about that for now.